src/classes/CampaignListSegmentMapper.cls

```
@@ -0,0 +1,131 @@
+     * name lookups
+     */
+    private static Map<Id, String> sourceNames = new Map<Id, String>();
+
+    /**
+     * @description Given the id of a source, i.e. a Report or Campaign, look
+     * up the name of that source.  This might execute a SOQL query.
+     *
+     * @param sourceId The id of the source (i.e. a Report id or a Campaign id
+     * @return String (the name of the source)
+     */
+    public String getSourceNameById(Id sourceId) {
+        if (!sourceNames.containsKey(sourceId)) {
+            String name;
+            try {
+                if (Report.sObjectType == sourceId.getSObjectType()) {
+                    name = 'Report: ' + [SELECT Name FROM Report WHERE Id = :sourceId].Name;
+                } else if (Campaign.sObjectType == sourceId.getSObjectType()) {
+                    name = 'Campaign: ' + [SELECT Name FROM Campaign WHERE Id = :sourceId].Name;
+                }
+            } catch (System.QueryException e) {
+                name = '';
+            }
+            sourceNames.put(sourceId, name);
+        }
+        return sourceNames.get(sourceId);
+    }
+
+    /**
+     * @description Given the id of the root segment of a campaign list, get the tree of segments that correspond to that campaign list.
+     *
+     * @param rootSegmentId The "campaign list" id
+     * @return CampaignListSegment
+     */
+    public CampaignList.Segment getSegmentTreeByRootId(Id rootSegmentId) {
+        CampaignList.Segment rootSegment;
+        Map<Id, CampaignList.Segment> segments = new Map<Id, CampaignList.Segment>();
+
+        for (Segment__c segment : queryByRootSegmentId(rootSegmentId)) {
+            segments.put(segment.Id, create(segment));
+        }
+
+        for (CampaignList.Segment segment : segments.values()) {
+            Id pid = segment.getParentId();
+
+            if (null == pid) {
+                rootSegment = segment;
+            } else if (segments.containsKey(pid)) {
```

Notes from code review with @davidhabib and @njjc:

In CampaignListSegmentMapper.getSegmentTreeByRootId(), if a queried Segment__c points to a parent with a different Root_Segment__c value, then this segment gets silently dropped.  While this is an indication of malformed data in the table, we should add something that detects this and throws an exception so that the malformed data can be dealt with, rather than silently ignoring it.